homework02/allan/string.cc

```diff
@@ -0,0 +1,103 @@
+#include "string.h"
+
+#include <cstring>
+#include <iostream>
+String::String(const char *ch)
+  : String(ch, strlen(ch) + 1)
+{
+}
+
+String::String()
+   :String("", 1)
+{
+}
+
+String::String(const String& other)
+  : String(other.c_str(), other.capacity())
+{
+}
+
+
+String::String(const char *ch, int capacity)
+  : mCapacity(capacity),
+    mData(std::make_unique<char []>(mCapacity))
+{
+  strcpy(mData.get(), ch);
+}
+
+String::~String()
+{
+}
+
+String& String::operator=(String other)
+{
+  std::swap(mCapacity, other.mCapacity);
+        std::swap(mData, other.mData);
+}
+
+bool String::empty() const
+{
+        return size() == 0;
+}
+
+const char* String::c_str() const
+{
+        return mData.get();
+}
+
+String::operator std::string() const
+{
+        return std::string(c_str());
+}
+
+char& String::operator[](int pos)
+{
+  return mData[pos];
+}
+
+const char& String::operator[](int pos) const
+{
+  return mData[pos];
+}
+
+int String::size() const
+{
+        return strlen(mData.get());
```

Comment:
`std::strlen`

Reply:
Yes a complexity promise is something I should think about. Goal for this cut was to reduce dependencies between members (size of buffer and size member)